### PR TITLE
fix: allow relative fetch to endpoint outside app from within handle

### LINKED ISSUE #9183: fetch inside a server hook throws error instead of forwarding request to proxied API
### Describe the bug

I have an app served behind a reverse proxy. Any call to `/api` gets forwarded to the API server, the others are forwarded to sveltekit.

Given an relative API endpoint `/api/endpoint`:

- Calling `event.fetch` inside a `load` or action handler works fine
- Calling `event.fetch` directly inside the `hooks.server.ts` gives a 404 error without making an HTTP request (I DON'T have any route on `/api/endpoint/+server.ts`, as it should be forwarded to the API server by the proxy) 

When calling `event.fetch` inside a `load` function I can see by logging `event.url.href` that the request to the same endpoint pass through `hooks.server.ts` just fine.

I think the issue is caused by fetch that shorts internal requests to their handler as said in the ([docs](https://kit.svelte.dev/docs/load#making-fetch-requests)):

> internal requests (e.g. for +server.js routes) go direct to the handler function when running on the server, without the overhead of an HTTP call.

It seems to do that only if `fetch` is called inside `hooks.server.ts`, should be the same behaviour regardless of where `fetch` is called?

As a solution maybe sveltekit should short-hand requests to an handler only if the route exists?

### Reproduction

Example:

- api.domain2.com - API server
- domain.com/api - Proxied API server
- domain.com - sveltekit app

Proxy configuration for development:

```typescript
// vite.config.ts

import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
  plugins: [sveltekit()],
  server: {
    proxy: {
      '/api': {
        target: 'http://api.domain2.com',
        changeOrigin: false,
        rewrite: (path) => path.replace(/^\/api/, '')
      }
    }
  }
});
```

```typescript
// src/hooks.server.ts

import type { Handle } from '@sveltejs/kit';

export const handle = (async ({ event, resolve }) => {
  console.log('Server Hook:', event.url.href);

  if (!event.locals.obj && event.url.pathname.startsWith('/private')) {
    try {
      const objResp = await event.fetch('/api/endpoint');
      if (!objResp.ok) throw objResp;
      event.locals.obj = await objResp.json();
    } catch {
      event.locals.obj = null;
    }
  }

  const response = await resolve(event);
  return response;
}) satisfies Handle;
```

This works fine:
```typescript
// src/private/+layout.server.ts

import type { LayoutServerLoad } from './$types';

export const load = (async ({ fetch }) => {

  const resp = await fetch('/api/endpoint');
        
  const obj = await resp.json();

  return { obj };
}) satisfies LayoutServerLoad;
```

### Logs

```Shell
Server Hook: http://localhost:5173/private
Server Hook: http://localhost:5173/api/endpoint
Error: Not found: /api/endpoint
    at resolve (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/respond.js:396:13)
    at resolve (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/respond.js:237:5)
    at Object.handle (/Users/damiano/Developer/dpdev/accounts/ui/src/hooks.server.ts:22:24)
    at Module.respond (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/respond.js:234:40)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async fetch (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:134:16)
    at async Object.eval [as fetch] (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:27:10)
    at async Object.handle (/Users/damiano/Developer/dpdev/accounts/ui/src/hooks.server.ts:14:29)
    at async Module.respond (/@fs/Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/runtime/server/respond.js:234:20)
    at async file:///Users/damiano/Developer/dpdev/accounts/node_modules/.pnpm/@sveltejs+kit@1.5.3_svelte@3.55.1+vite@4.1.1/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:495:22
```

### System Info

```Shell
System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 97.48 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 19.6.1 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.4.0 - /opt/homebrew/bin/npm
  Browsers:
    Firefox Developer Edition: 110.0
    Safari: 16.2
```

### Severity

serious, but I can work around it

### Additional Information

As a workaround you can define a route that calls fetch directly:
```typescript
// src/workaround/[...path]/+server.ts
import type { RequestHandler } from './$types';

export const GET = (({ params, url }) => {
  return fetch(`/${params.path + url.search}`);
}) satisfies RequestHandler;
```

```typescript
// src/hooks.server.ts

import type { Handle } from '@sveltejs/kit';

export const handle = (async ({ event, resolve }) => {
  console.log('Server Hook:', event.url.href);

  if (!event.locals.obj && event.url.pathname.startsWith('/private')) {
    try {
      const objResp = await event.fetch('/workaround/api/endpoint');
      if (!objResp.ok) throw objResp;
      event.locals.obj = await objResp.json();
    } catch {
      event.locals.obj = null;
    }
  }

  const response = await resolve(event);
  return response;
}) satisfies Handle;
```

### PR DESCRIPTION
**Pull request by dummdidumm**
fixes #9183
also consolidates usage of state.initiator across the server, only setting it in one place now where it's correct, prior to calling the internal fetch function. this means it now compares by url string, not matched route

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: packages/kit/src/runtime/server/page/index.js [hunk 1]
```diff
@@ -18,24 +18,21 @@ import { get_data_json } from '../data/index.js';
...
  * @param {import('types').SSROptions} options
  * @param {import('types').SSRManifest} manifest
  * @param {import('types').SSRState} state
  * @param {import('types').RequiredResolveOptions} resolve_opts
  * @returns {Promise<Response>}
  */
-export async function render_page(event, route, page, options, manifest, state, resolve_opts) {
-	if (state.initiator === route) {
+export async function render_page(event, page, options, manifest, state, resolve_opts) {
+	if (state.initiator === event.request.url) {
```
Comment (dummdidumm) [inline packages/kit/src/runtime/server/page/index.js:29]: This now compares by url, not matched route anymore. Does this have any buggy implications that I'm not aware of?

That infinite loop check is a bit too simple anyway, if you have a loop that spans more than one fetch we don't catch it right now - but it also doesn't seem to have happened to anyone yet, so something to worry about another day I'd say.

File: packages/kit/types/internal.d.ts [hunk 2]
```diff
@@ -157,15 +157,6 @@ export type RecursiveRequired<T> = {
 
 export type RequiredResolveOptions = Required<ResolveOptions>;
 
-export interface Respond {
```
Comment (dummdidumm) [inline packages/kit/types/internal.d.ts:160]: inlined this into the function because it was annoying that "go to definition" showed two positions (the type and the implementation) when I just want to go to the implementation

File: .changeset/good-bikes-cross.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+fix: allow relative fetch to enpoint outside app from within handle
```
Comment (benmccann) [inline .changeset/good-bikes-cross.md:5]:
```suggestion
fix: allow relative fetch to endpoint outside app from within handle
```

File: packages/kit/types/internal.d.ts [hunk 4]
```diff
@@ -380,7 +367,7 @@ export interface SSRRoute {
 export interface SSRState {
 	fallback?: string;
 	getClientAddress(): string;
-	initiator?: SSRRoute | SSRErrorPage;
+	initiator?: 'GENERIC_ERROR' | (string & {});
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:370]: what does `string & {}` mean?

File: packages/kit/src/runtime/server/page/index.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/index.js:29]: Not opposed to this change but what's the reason for it?

**Review by Rich-Harris — APPROVED** [review]

File: packages/kit/src/runtime/server/page/index.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/index.js:29]: ah nvm, i see the explanation in the original comment. i honestly don't know if it's potentially buggy

File: packages/kit/types/internal.d.ts (hunk 4 above)
Comment (dummdidumm) [reply packages/kit/types/internal.d.ts:370]: It's a trick so that you still get autocompletion for the special string. If you just do `| string` TS would widen the whole type to just `string`

File: packages/kit/src/runtime/server/fetch.js [hunk 5]
```diff
@@ -131,6 +131,7 @@ export function create_fetch({ event, options, manifest, state, get_cookie_heade
 					);
 				}
 
+				state.initiator = request.url;
```
Comment (dummdidumm) [inline packages/kit/src/runtime/server/fetch.js:134]:
```suggestion
				state.initiator = event.request.url;
```

File: packages/kit/src/runtime/server/page/index.js (hunk 1 above)
Comment (dummdidumm) [reply packages/kit/src/runtime/server/page/index.js:29]: Also this doesn't catch redirects right? I'm wondering at this point if this just is some ancient guard we no longer need.

File: packages/kit/types/internal.d.ts (hunk 4 above)
Comment (Rich-Harris) [reply packages/kit/types/internal.d.ts:370]: 🤯

File: packages/kit/src/runtime/server/page/index.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/index.js:29]: We do need it, otherwise there's no way to fall back to `fetch` in the case where `event.fetch` is used to request something other than a route in the currently available manifest, no way to bail out in the case where the root layout fetches a non-existent resource, and no way to bail out when a route fetches itself. If we get rid of it a bunch of tests start failing.

Thinking further, using the URL rather than the `route` object does make me nervous. It fails in these edge cases, resulting in an infinite loop:

* `${url.pathname}?random=${Math.random()}` (would be fixed by using `event.url.pathname`)
* `${url.pathname}x` when matched against `/[...catchall]` (would not)

These might be vanishingly unlikely but IIRC the `initiator` stuff was added because we encountered a similarly vanishingly unlikely scenario in the wild. Infinite request loops can result in huge compute bills (I've seen this happen!) so we should strenuously avoid them where we can

File: packages/kit/src/runtime/server/fetch.js [hunk 6]
```diff
@@ -131,6 +131,7 @@ export function create_fetch({ event, options, manifest, state, get_cookie_heade
 					);
 				}
 
+				state.initiator = event.request.url;
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/fetch.js:134]: putting it here undoes this line, causing an infinite loop in the case where `event.fetch` happens in a root layout while rendering an error page https://github.com/sveltejs/kit/blob/80a15056d3e7d36ec35bb4339afe3e5c5d8d45f4/packages/kit/src/runtime/server/page/respond_with_error.js#L46

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
Ah, ok, I finally understand why this uses `request.url` — if it's happening in a `handle` hook there's no matched route.

We could try matching against the route _and_ the pathname, but it still wouldn't handle a case where `/a` fetches `/b` and `/b` fetches `/a`. Might be simpler to just track the request 'depth', same as we track the length of redirect chains

File: packages/kit/src/runtime/server/fetch.js (hunk 6 above)
Comment (dummdidumm) [reply packages/kit/src/runtime/server/fetch.js]: Good catch, this needs a `if not generic error initiator ` check then

**Comment by dummdidumm** [comment]
Another idea would be to track the urls that are fetched and bail if we encounter a url we already saw. But that maybe also needs to account for headers and the body (who knows...), so yeah, it's probably better to just track the depth. 10 sounds sensible, it's the same as in the client
